rstar/src/algorithm/selection_functions.rs

```diff
@@ -37,6 +37,22 @@ where
     }
 }
 
+/// Similar to [`SelectionFunction`](trait.SelectionFunction.html) but additional data can be returned by
+/// `should_unpack_leaf`.
+pub trait SelectionFunctionWithData<T, D>
+where
+    T: RTreeObject,
+{
```

Comment:
Would it be simpler to simply pass the `RTreeNode` directly?  The assumption is that root is implicitly unpacked as it's just a `Vec<RTreeNode>`.  Then the trait just needs one method and we don't have to force that the decision is made based on envelope alone.

@expenses @adamreichold @urschrei

Reply:
This does sound sensible on a first reading, but I think such a change would be orthogonal to what is proposed here. (It would apply to the original `SelectionFunction` trait just as well.) Hence I would suggest not making this PR any more complicated than it already is and tackling this separately.

Reply:
Apologies, I didn't realize this is an extension of the existing `SelectionFunction` trait.  But, now I'm more confused.  Given the data `D` is being computed during the `should_unpack_leaf` call,  can't it just be computed during the iteration using the vanilla `locate_with_selection_function`?   What's the extra use-case being handled by this new trait?  Sorry if I'm missing something obvious here.

Reply:
I'd say it is an "API courtesy". While you could have a mutable binding initialised to `None` and set to `Some` within the selection function before it returns true and then use this in downstream code, that is somewhat contrived, possibly less efficient and really hard to read when iterator adaptors are chained onto the selection method call.

Reply:
Is there a good example which could describe this use case further? I'm wondering how this could be documented, partially because I still don't understand what this could be used for.

Also, if this is about information calculated *within* the selection function, shouldn't the signatures of `SelectionFunctionWithData` rather be `&mut self` (instead of `&self`)?

Reply:
The original poster mentioned

> This is useful for things like ray intersections, where you want to return the time of impact in order to find the closest object.

in the cover letter.

> Also, if this is about information calculated within the selection function, shouldn't the signatures of SelectionFunctionWithData rather be &mut self (instead of &self)?

I don't think that computing a new value and returning that necessarily requires modifying state just as computing a `bool` indicating whether to unpack or not does not require it.

But then if the implementation can provide an exclusive reference, that would provide a strictly more general API by offering both choices of `&self` and `&mut self` to downstream code and would thereby seem to be preferable. But this line of reasoning would just as well apply to the existing `SelectionFunction` trait again.